Fix checking if newline is needed before `else` in let-else statement

rhysd opened the pull request:
Fixes #5901

ytmimi commented:
@rhysd I haven't had a chance to review this yet, but I wanted to see how your changes handle this case:

```rust
fn issue5901() {
    #[cfg(target_os = "linux")]
    // some comment between attr and let-else
    let Some(x) = foo() else { return; };
}

fn issue5901() {
    #[cfg(target_os = "linux")]
    /* some comment between attr and let-else */
    let Some(x) = foo() else { return; };
}
```

rhysd commented:
Thank you for the point. I'll try to add new test.

rhysd commented:
I confirmed that this fix didn't work for the comment between the attributes and let-else statement. I'll have a look whether it can be fixed.

```diff
 fn main() {
     #[cfg(target_os = "linux")]
     // this is test
-    let x = 42 else { todo!() };
+    let x = 42
+    else {
+        todo!()
+    };
 }
```

ytmimi commented:
@rhysd thanks for taking a look 🙏🏼. Before you keep working on this I also want to point out that [any change needs to be version gated](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#version-gate-formatting-changes) since let-else support was released as stable and we can't break our formatting guarantees.

rhysd commented:
@ytmimi Thanks for the information. I'll check the documentation. And also, I may find a way to strip things before `let` correctly. I'll update this branch once I add more tests and the version gate.

rhysd commented:
@ytmimi I force-pushed the fix. I added more tests and v1 version gate as well. Is there any place to add tests for v1 formatting? Currently only v2 formatting is tested.

ytmimi commented:
You should create a `let_else_v2.rs` test for this. See the [Create test cases section of the Contributing guide](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#create-test-cases) for details on how to add configuration values to tests via comments.

ytmimi commented on src/items.rs:
```diff
@@ -127,8 +128,15 @@ impl Rewrite for ast::Local {
                 let else_kw_span = init.span.between(block.span);
+                // Strip attributes and comments to check if newline is needed before the else
+                // keyword from the initializer part. (#5901)
+                let init_str = if context.config.version() == Version::One {
```
Can we invert the logic to use `Version::Two` instead of `Version::One`.

ytmimi commented on src/items.rs:
```diff
@@ -127,8 +128,15 @@ impl Rewrite for ast::Local {
+                };
                 let force_newline_else = pat_str.contains('\n')
-                    || !same_line_else_kw_and_brace(&result, context, else_kw_span, nested_shape);
+                    || !same_line_else_kw_and_brace(init_str, context, else_kw_span, nested_shape);
```
There are two calls bellow that I think we need to change as well.

The `available_space` calculation should also be derived using the len without any leading attributes or comments. The call to `allow_single_line_let_else_block` should also take a string starting at the `let` kw.

ytmimi commented on tests/source/let_else.rs:
```diff
@@ -160,3 +160,16 @@ fn with_trailing_try_operator() {
     let Ok(Some(next_bucket)) = ranking_rules[cur_ranking_rule_index].next_bucket(ctx, logger, &ranking_rule_universes[cur_ranking_rule_index]) else { return };
 }
+
+fn issue5901() {
+    #[cfg(target_os = "linux")]
+    let Some(x) = foo() else { return; };
+
+    #[cfg(target_os = "linux")]
+    // Some comments between attributes and let-else statement
+    let Some(x) = foo() else { return; };
+
+    #[cfg(target_os = "linux")]
+    #[cfg(target_arch = "x86_64")]
+    let Some(x) = foo() else { return; };
+}
```
As mentioned, you can move these tests to a `let_else_v2.rs` test file.

ytmimi reviewed (commented):
Thanks again for taking the time to look into this. I think you're on the right track with getting the `let_kw_offset` early on. As mentioned, there are two other calls that use the result below for their calculations and those need to be changed as well.

rhysd commented:
@ytmimi Thank you for your review comments. I updated this branch to resolve them.

- Preferred `Version::Two` to `Version::One`
- Added test cases for both v1 and v2 formatting
- Fixed calculating available space stripping things before `let`

ytmimi commented on src/items.rs:
```diff
@@ -146,11 +154,21 @@ impl Rewrite for ast::Local {
                     std::cmp::min(shape.width, context.config.single_line_let_else_max_width());
 
                 // If available_space hits zero we know for sure this will be a multi-lined block
-                let available_space = max_width.saturating_sub(result.len());
+                let assign_len = if context.config.version() == Version::Two {
+                    result.len() - let_kw_offset
+                } else {
+                    result.len()
+                };
+                let available_space = max_width.saturating_sub(assign_len);
 
+                let assign_str = if context.config.version() == Version::Two {
+                    &result[let_kw_offset..]
+                } else {
+                    result.as_str()
+                };
                 let allow_single_line = !force_newline_else
                     && available_space > 0
-                    && allow_single_line_let_else_block(&result, block);
+                    && allow_single_line_let_else_block(assign_str, block);
```
Thanks for looking into this change. I think it's simpler to rewrite this as follows:

```rust
    let assign_str_with_else_kw = if context.config.version() == Version::Two {
        &result[let_kw_offset..]
    } else {
        result.as_str()
    };

    let available_space = max_width.saturating_sub(assign_str_with_else_kw.len());

    let allow_single_line = !force_newline_else
        && available_space > 0
        && allow_single_line_let_else_block(assign_str_with_else_kw, block);
```

rhysd replied:
That's simpler. Thank you for the suggestion :+1:

ytmimi commented on tests/source/let_else.rs:
```diff
@@ -160,3 +160,23 @@ fn with_trailing_try_operator() {
+
+fn issue5901() {
+    #[cfg(target_os = "linux")]
+    let Some(x) = foo else { return; };
```
I appreciate you adding both v1 and v2 test cases for this.

Is there a reason why these test cases use `return;` instead of `todo!()` like the last one?

Because `return;` is a statement rustfmt will always rewrite the else block over multiple lines in those cases.


Also, I'd like to increase the coverage with attributes. When you get a chance please take some of the longer test cases from the v1 test file and add attributes to them and incorporate them in the v2 test file (the more the better 😁).

rhysd replied:
> Is there a reason why these test cases use return; instead of todo!() like the last one?

Actually there is no reason. I'll replace `return`s with `todo!`.

> When you get a chance please take some of the longer test cases from the v1 test file and add attributes to them and incorporate them in the v2 test file

Sure. I'll try to do that.

rhysd commented:
@ytmimi Thank you for your review comments again. I addressed them.

- Simplified checking if the else block can be single-lined
- Preferred `todo!()` expression to `return;` statement
- Imported more v1 test cases into `let_else_v2.rs`

ytmimi reviewed (approved):
Thanks for helping to fix this issue and for your first contribution to rustfmt 🥳

ytmimi commented:
As a last step I want to run our [diff-check job](https://github.com/rust-lang/rustfmt/actions/runs/6126768531), It should pass since the changes are version gated, but I still want to go through this process.
